### ISSUE TITLE
master branch of openshift-ansible fails to deploy the OpenShift registry on OSE 3.2

### ISSUE THREAD

**Issue by lhuard1A** [openshift/openshift-ansible#4000]
#### Description

`openshift-ansible` should be able to deploy any major version of OpenShift since the major version of OpenShift is selected by the channel which is subscribed to.

However, deploying an OSE 3.2 cluster with the current `master` version of openshift-ansible fails.

##### Version

* OSE 3.2
* openshift-ansible master

##### Steps To Reproduce

1. Subscribe the machines to rhel-7-server-ose-3.2-rpms channel
2. Deploy an OpenShift cluster with the `master` version of openshift-ansible

##### Expected Results

The deployment completes and openshift-ansible exits without any error.

##### Observed Results

openshift-ansible fails at the followinq task:
```
TASK [openshift_hosted : Create OpenShift registry] ****************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: __main__.RegistryException: Could not perform registry preparation. {'returncode': 1, 'stderr': u'Error: unknown flag: --daemonset\n\nUsage:\n  oc adm registry [options]\n\nExamples:\n  # Check if default Docker registry ("docker-registry") has been created\n  $ oc adm registry --dry-run\n\n  # See what the registry will look like if created\n  $ oc adm registry -o yaml\n\n  # Create a registry with two replicas if it does not exist\n  $ oc adm registry --replicas=2\n\n  # Use a different registry image\n  $ oc adm registry --images=myrepo/docker-registry:mytag\n\nOptions:\n      --create=false: deprecated; this is now the default behavior\n      --credentials=\'\': Path to a .kubeconfig file that will contain the credentials the registry should use to contact the master.\n      --dry-run=false: Check if the registry exists instead of creating.\n      --images=\'openshift3/ose-${component}:${version}\': The image to base this registry on - ${component} will be replaced with --type\n      --labels=\'docker-registry=default\': A set of labels to uniquely identify the registry and its components.\n      --latest-images=false: If true, attempt to use the latest image for the registry instead of the latest release.\n      --mount-host=\'\': If set, the registry volume will be created as a host-mount at this path.\n      --no-headers=false: When using the default output, don\'t print headers.\n  -o, --output=\'\': Output format. One of: json|yaml|wide|name|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=... See golang template [http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template [http://releases.k8s.io/release-1.2/docs/user-guide/jsonpath.md].\n      --output-version=\'\': Output the formatted object with the given group version (for ex: \'extensions/v1beta1\').\n      --ports=\'5000\': A comma delimited list of ports or port pairs to expose on the registry pod. The default is set for 5000.\n      --replicas=1: The replication factor of the registry; commonly 2 when high availability is desired.\n      --selector=\'\': Selector used to filter nodes on deployment. Used to run registries on a specific set of nodes.\n      --service-account=\'registry\': Name of the service account to use to run the registry pod.\n  -a, --show-all=true: When printing, show all resources (false means hide terminated pods.)\n      --show-labels=false: When printing, show all labels as the last column (default hide labels column)\n      --sort-by=\'\': If non-empty, sort list types using this field specification.  The field specification is expressed as a JSONPath expression (e.g. \'{.metadata.name}\'). The field in the API resource specified by this JSONPath expression must be an integer or a string.\n  -t, --template=\'\': Template string or path to template file to use when -o=go-template, -o=go-template-file. The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].\n      --tls-certificate=\'\': An optional path to a PEM encoded certificate (which may contain the private key) for serving over TLS\n      --tls-key=\'\': An optional path to a PEM encoded private key for serving over TLS\n      --type=\'docker-registry\': The registry image to use - if you specify --images this flag may be ignored.\n      --volume=\'/registry\': The volume path to use for registry storage; defaults to /registry which is the default for origin-docker-registry.\n\nUse "oc adm options" for a list of global command-line options (applies to all commands).\n\n', 'stdout': u'', 'cmd': '/bin/oc adm registry --daemonset=False --enforce-quota=False --images=openshift3/ose-${component}:${version} --latest-images=False --ports=5000 --replicas=1 --selector=type=infra --service-account=registry --dry-run=True -o json -n default', 'results': {}}
fatal: [lenaic-master-ebea6]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_8Ri5NU/ansible_module_oc_adm_registry.py\", line 2670, in <module>\n    main()\n  File \"/tmp/ansible_8Ri5NU/ansible_module_oc_adm_registry.py\", line 2662, in main\n    results = Registry.run_ansible(module.params, module.check_mode)\n  File \"/tmp/ansible_8Ri5NU/ansible_module_oc_adm_registry.py\", line 2598, in run_ansible\n    api_rval = ocregistry.create()\n  File \"/tmp/ansible_8Ri5NU/ansible_module_oc_adm_registry.py\", line 2399, in create\n    self.needs_update()\n  File \"/tmp/ansible_8Ri5NU/ansible_module_oc_adm_registry.py\", line 2500, in needs_update\n    not Utils.check_def_equal(self.prepared_registry['service'].yaml_dict,\n  File \"/tmp/ansible_8Ri5NU/ansible_module_oc_adm_registry.py\", line 2289, in prepared_registry\n    results = self.prepare_registry()\n  File \"/tmp/ansible_8Ri5NU/ansible_module_oc_adm_registry.py\", line 2357, in prepare_registry\n    raise RegistryException('Could not perform registry preparation. {}'.format(results))\n__main__.RegistryException: Could not perform registry preparation. {'returncode': 1, 'stderr': u'Error: unknown flag: --daemonset\\n\\nUsage:\\n  oc adm registry [options]\\n\\nExamples:\\n  # Check if default Docker registry (\"docker-registry\") has been created\\n  $ oc adm registry --dry-run\\n\\n  # See what the registry will look like if created\\n  $ oc adm registry -o yaml\\n\\n  # Create a registry with two replicas if it does not exist\\n  $ oc adm registry --replicas=2\\n\\n  # Use a different registry image\\n  $ oc adm registry --images=myrepo/docker-registry:mytag\\n\\nOptions:\\n      --create=false: deprecated; this is now the default behavior\\n      --credentials=\\'\\': Path to a .kubeconfig file that will contain the credentials the registry should use to contact the master.\\n      --dry-run=false: Check if the registry exists instead of creating.\\n      --images=\\'openshift3/ose-${component}:${version}\\': The image to base this registry on - ${component} will be replaced with --type\\n      --labels=\\'docker-registry=default\\': A set of labels to uniquely identify the registry and its components.\\n      --latest-images=false: If true, attempt to use the latest image for the registry instead of the latest release.\\n      --mount-host=\\'\\': If set, the registry volume will be created as a host-mount at this path.\\n      --no-headers=false: When using the default output, don\\'t print headers.\\n  -o, --output=\\'\\': Output format. One of: json|yaml|wide|name|go-template=...|go-template-file=...|jsonpath=...|jsonpath-file=... See golang template [http://golang.org/pkg/text/template/#pkg-overview] and jsonpath template [http://releases.k8s.io/release-1.2/docs/user-guide/jsonpath.md].\\n      --output-version=\\'\\': Output the formatted object with the given group version (for ex: \\'extensions/v1beta1\\').\\n      --ports=\\'5000\\': A comma delimited list of ports or port pairs to expose on the registry pod. The default is set for 5000.\\n      --replicas=1: The replication factor of the registry; commonly 2 when high availability is desired.\\n      --selector=\\'\\': Selector used to filter nodes on deployment. Used to run registries on a specific set of nodes.\\n      --service-account=\\'registry\\': Name of the service account to use to run the registry pod.\\n  -a, --show-all=true: When printing, show all resources (false means hide terminated pods.)\\n      --show-labels=false: When printing, show all labels as the last column (default hide labels column)\\n      --sort-by=\\'\\': If non-empty, sort list types using this field specification.  The field specification is expressed as a JSONPath expression (e.g. \\'{.metadata.name}\\'). The field in the API resource specified by this JSONPath expression must be an integer or a string.\\n  -t, --template=\\'\\': Template string or path to template file to use when -o=go-template, -o=go-template-file. The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].\\n      --tls-certificate=\\'\\': An optional path to a PEM encoded certificate (which may contain the private key) for serving over TLS\\n      --tls-key=\\'\\': An optional path to a PEM encoded private key for serving over TLS\\n      --type=\\'docker-registry\\': The registry image to use - if you specify --images this flag may be ignored.\\n      --volume=\\'/registry\\': The volume path to use for registry storage; defaults to /registry which is the default for origin-docker-registry.\\n\\nUse \"oc adm options\" for a list of global command-line options (applies to all commands).\\n\\n', 'stdout': u'', 'cmd': '/bin/oc adm registry --daemonset=False --enforce-quota=False --images=openshift3/ose-${component}:${version} --latest-images=False --ports=5000 --replicas=1 --selector=type=infra --service-account=registry --dry-run=True -o json -n default', 'results': {}}\n", "module_stdout": "", "msg": "MODULE FAILURE"}
```

##### Additional Information

The reason of the above error is:
```
Error: unknown flag: --daemonset
```

The point is that `oc adm registry` didn’t have the `--daemonset` in OSE 3.2. It learned that flag in more recent versions.

**Comment by sdodson** [comment]
@lhuard1A For quite some time we've only supported the current development (3.6) and released (3.5) versions of openshift from the master branch. We've discussed moving back to a model where we support more versions from master branch but that would not be retroactive once we decide to do that it'd just be forward looking, ie: master would support 3.5, 3.6, and 3.7 where as you'd still need to use release-1.2, release-1.3, release-1.4 branches to install 3.2, 3.3, and 3.4 respectively.

**Comment by lhuard1A** [comment]
Thanks @sdodson for your prompt answer!
I had a deeper look at this precise issue and figured out that the only problem was that openshift-ansible tried to pass to `oc adm registry` a flag which was not known.
As the default value of this parameter in oc 3.5 is false and as the value that is passed by openshift-ansible by default is false, it should be safe to not pass that flag when the default is false.

Hence my PR #4003.

Once that issue fixed, I managed to deploy an OSE 3.2 cluster with the master branch of openshift-ansible.
So, I’m wondering what is implied by “only the dev (3,6) and current (3.5) OCP versions are supported in master”?
Does it only mean that you don’t validate that it works and you wouldn’t block a PR that breaks backward compatibility? Or are there some features which are known to be broken when deploying an old version of OpenShift with master of openshift-ansible?

**Comment by sdodson** [comment]
We (Red Hat) definitely make no effort to test master on releases older than 3.5. We have and will remove some amount of conditional logic if it makes the code cleaner and easier to maintain going forward. Also, I think the current logging and metrics playbooks likely don't work on anything older than 3.4.

I'm fine with accepting fixes so long as it doesn't needlessly complicate the codebase but just know we're not actively testing old releases with master so there's probably more problems we're unaware of.
